WIP: Parse Apple SDK versions

madsmtm opened the pull request:
High-level goals of this PR:

1. Easy cross compile on macOS host to iOS, tvOS, watchOS and visionOS.
    - If SDKROOT is specified for such a target, we need to find a macOS SDK to compile build scripts etc. for the host.
2. Support cross compile to older macOS targets.
    - Similar to above, we are potentially given an SDK that is not suitable for the host machine (e.g. Aarch64 to x86 with the macOS 10.13 SDK that does not support Aarch64).
3. Support cross compile on non-macOS host.
    - Give a warning if no SDK root is set, and guide the user towards setting it.
    - Should still support `zig cc` (though that maybe only works on macOS? I don't think Zig provides the `.tdb`s and headers to build and link iOS?)

---

Previously, we did a simplistic check against the `SDKROOT` path to try to figure out whether we should use the environment variable, or try to find a better matching SDK instead.

This is brittle since the user might want their SDK to be in a different place in the file system (that's kinda the whole idea of the `SDKROOT` variable), and also didn't work properly with SDKs in the Xcode Command Line Tools.

Instead, we now:
1. Parse the `SDKSettings.json` file under the `SDKROOT`, to determine whether it is suitable for the target we're compiling for (which it might not be when e.g. using Cargo, and targetting iOS but compiling build scripts on macOS).
2. If it's not, attempt to find another SDK.

This allows us to give much more detailed error messages, to better support cross-compile scenarios, and allows using parameters from the SDK, such as the SDK version, which we need to pass to `ld64` for correctness (this fixes the remaining difference between linking with `cc` and `ld64`).

Final part of https://github.com/rust-lang/rust/issues/129432.

Additionally, this fixes https://github.com/rust-lang/rust/issues/80120 by introducing better error messages, so that it's possible to more clearly see when `/Library/Developer/CommandLineTools` is invalid.

### Test

Tested with:

```
./x test tests/run-make/apple-* --target="aarch64-apple-darwin,aarch64-apple-ios,aarch64-apple-ios-macabi,aarch64-apple-ios-sim,aarch64-apple-tvos,aarch64-apple-tvos-sim,aarch64-apple-visionos,aarch64-apple-visionos-sim,aarch64-apple-watchos,aarch64-apple-watchos-sim,arm64_32-apple-watchos,armv7k-apple-watchos,armv7s-apple-ios,x86_64-apple-darwin,x86_64-apple-ios,x86_64-apple-ios-macabi,x86_64-apple-tvos,x86_64-apple-watchos-sim,x86_64h-apple-darwin"
IPHONEOS_DEPLOYMENT_TARGET=10.0 ./x test tests/run-make/apple-* --target=i386-apple-ios
MACOSX_DEPLOYMENT_TARGET=10.12 SDKROOT=$(pwd)/MacOSX10.13.sdk ./x test tests/run-make/apple-* --target=i686-apple-darwin
```

Which includes every Apple target [except for arm64e](https://github.com/rust-lang/rust/issues/130085).

Note that using the `MacOSX10.13.sdk` now works (and should work even more reliably once this change hits beta).


### TODO

- [ ] Manually test a few more scenarios:
  - [x] Using `cargo +stage1` inside Xcode for a project with both build scripts and proc macros (namely [Bevy's](https://github.com/bevyengine/bevy/tree/main/examples/mobile)).
  - [ ] Use `zig cc` without any SDKs available.
  - [ ] Cross compile with `zig cc` from a non macOS host.
  - [ ] Set SDKROOT to different crazy values.
  - [ ] Test running `xcrun --sdk XYZ rustc +stage1`.
- [ ] Clean up diagnostic setup (and use subdiagnostics for certain things?).
- [ ] Fix bootstrap so that it doesn't pass `SDKROOT` by default
  - Not sure if this is possible, since we also want the user to be able to override SDKROOT when building with bootstrap?
- [ ] Test properly with i686

CC @BlackHoleFox @thomcc

Currently an unholy `git` combination of https://github.com/rust-lang/rust/pull/131037, https://github.com/rust-lang/rust/pull/131433 and https://github.com/rust-lang/rust/pull/131477, so:
@rustbot blocked

madsmtm committed:
Emit diagnostics for incorrect deployment targets

madsmtm committed:
Move versioned LLVM target creation to rustc_codegen_ssa

The OS version depends on the deployment target environment variables,
the access of which we want to move to later in the compilation pipeline
that has access to more information, for example `env_depinfo`.

madsmtm committed:
Re-implement SDK discovery instead of using xcrun

madsmtm committed:
Move knowledge of SDK names to rustc_target

Also make the SDK name be the same casing as used in the file system.

madsmtm committed:
Merge branch 'move-llvm-target-versioning' into sdkroot

madsmtm committed:
Pass SDK version to ld64 when linking

This is an important property of the binary that we were previously
setting incorrectly, and is a big step towards making binaries linked
with cc and ld64 equal.

madsmtm committed:
Refactor ld64 arch determination

madsmtm committed:
Always attempt to get SDK root, and pass it to Clang via env var

The exact reasoning why we do not always pass the SDK root with
`-isysroot` to `cc` when linking on macOS eludes me (the git history
dead ends in #100286), but I suspect it's because we want to support
`cc`s which do not support this option.

So instead, we pass the SDK root via the environment variable SDKROOT.
This way, compiler drivers that support setting the SDK root (such as
Clang and GCC) can use it, while compiler drivers that don't (presumably
because they figure out the SDK root in some other way) can just ignore
it.

This fixes https://github.com/rust-lang/rust/issues/80817 (by always
passing the SDK root, even when linking with cc on macOS).

madsmtm committed:
Convert OSVersion to a struct that implements traits

Instead of being a type-alias to a tuple. This allows us to implement
Deserialize for OSVersion, which will be useful when parsing SDK
settings.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11265701411/job/31327973602) [(plain)](https://github.com/rust-lang/rust/commit/9e37d1b81332f85d296ebc2c95e249fa8a53cb89/checks/31327973602/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.704 Building wheels for collected packages: reuse
#13 2.705   Building wheel for reuse (pyproject.toml): started
#13 2.949   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 2.950   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 2.950   Stored in directory: /tmp/pip-ephem-wheel-cache-_g6z7v9x/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 2.953 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.340 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.340 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.873 Collecting virtualenv
#13 3.873 Collecting virtualenv
#13 3.914   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 3.989      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 82.6 MB/s eta 0:00:00
#13 4.048 Collecting filelock<4,>=3.12.2
#13 4.053   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.086 Collecting platformdirs<5,>=3.9.1
#13 4.090   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.110 Collecting distlib<1,>=0.3.7
#13 4.122      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 81.8 MB/s eta 0:00:00
#13 4.122      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 81.8 MB/s eta 0:00:00
#13 4.202 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.389 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 4.5s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      192448 kB
DirectMap2M:     9244672 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
fmt check
Diff in /checkout/compiler/rustc_codegen_ssa/src/apple.rs:383:
     /// This is needed by the linker.
     ///
     /// settings["Version"] or settings["VersionMap"]["macOS_iOSMac"][settings["Version"]].
-    pub(crate) fn sdk_version(&self, target: &Target, sdkroot: &Path) -> Result<AppleOSVersion, AppleSdkError> {
+    pub(crate) fn sdk_version(
+        target: &Target,
+        sdkroot: &Path,
+        sdkroot: &Path,
+    ) -> Result<AppleOSVersion, AppleSdkError> {
         if target.abi == "macabi" {
                 .version_map
                 .version_map
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_ssa/src/common.rs" "/checkout/compiler/rustc_codegen_ssa/src/mono_item.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/coverageinfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/locals.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/mod.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/rvalue.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/statement.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/analyze.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/intrinsic.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/debuginfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/operand.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/place.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/constant.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/block.rs" "/checkout/compiler/rustc_codegen_ssa/src/codegen_attrs.rs" "/checkout/compiler/rustc_codegen_ssa/src/size_of_val.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/coverageinfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/misc.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/abi.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/mod.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/type_.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/backend.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/builder.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/declare.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/asm.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/intrinsic.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/debuginfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/statics.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/consts.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/write.rs" "/checkout/compiler/rustc_codegen_ssa/src/errors.rs" "/checkout/compiler/rustc_codegen_ssa/src/apple.rs" "/checkout/compiler/rustc_codegen_ssa/src/meth.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/link.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/archive.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/mod.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/rpath/tests.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/metadata.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/rpath.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/command.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/symbol_export.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/lto.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/write.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/linker.rs" "/checkout/compiler/rustc_codegen_ssa/src/target_features.rs" "/checkout/compiler/rustc_codegen_ssa/src/lib.rs" "/checkout/compiler/rustc_mir_build/src/check_unsafety.rs" "/checkout/compiler/rustc_mir_build/src/build/coverageinfo.rs" "/checkout/compiler/rustc_mir_build/src/build/misc.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/as_operand.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/mod.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/as_temp.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/as_constant.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/into.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/as_place.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/as_rvalue.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/stmt.rs" "/checkout/compiler/rustc_mir_build/src/build/expr/category.rs" "/checkout/compiler/rustc_mir_build/src/build/mod.rs" "/checkout/compiler/rustc_mir_build/src/build/matches/match_pair.rs" "/checkout/compiler/rustc_mir_build/src/build/matches/mod.rs" "/checkout/compiler/rustc_mir_build/src/build/matches/test.rs" "/checkout/compiler/rustc_mir_build/src/build/matches/util.rs" "/checkout/compiler/rustc_codegen_ssa/src/apple/tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Oct 10 01:25:18 UTC 2024
  network time: Thu, 10 Oct 2024 01:25:18 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

madsmtm committed:
Parse SDK settings

Previously, we did a simplistic check against the SDKROOT path to try
to figure out whether we should use the environment variable, or try to
find a better matching SDK instead.

This is brittle since the user might want their SDK to be in a
different place in the file system (that's kinda the whole idea of the
SDKROOT variable), and also didn't work properly with SDKs in the Xcode
Command Line Tools.

Instead, we now:
1. Parse the SDKSettings.json file under the SDKROOT, to determine
   whether it is suitable for the target we're compiling for (which it
   might not be when using Cargo, and targetting iOS but compiling build
   scripts on macOS).
2. If it's not, attempt to find another SDK.

This allows us to give much more detailed error messages, to better
support cross-compile scenarios, and allows using parameters from the
SDK (such as the SDK version) elsewhere in the future.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11265810143/job/31328276242) [(plain)](https://github.com/rust-lang/rust/commit/1960f3b302472ba520ec87f164a85b365c78716f/checks/31328276242/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.789 Building wheels for collected packages: reuse
#13 2.790   Building wheel for reuse (pyproject.toml): started
#13 3.036   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.037   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.037   Stored in directory: /tmp/pip-ephem-wheel-cache-z57sj4k1/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.040 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.429 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.430 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.943 Collecting virtualenv
#13 3.943 Collecting virtualenv
#13 4.003   Downloading virtualenv-20.26.6-py3-none-any.whl (6.0 MB)
#13 4.294      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 20.8 MB/s eta 0:00:00
#13 4.352 Collecting filelock<4,>=3.12.2
#13 4.362   Downloading filelock-3.16.1-py3-none-any.whl (16 kB)
#13 4.394 Collecting platformdirs<5,>=3.9.1
#13 4.403   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.425 Collecting distlib<1,>=0.3.7
#13 4.447      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 45.4 MB/s eta 0:00:00
#13 4.447      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 45.4 MB/s eta 0:00:00
#13 4.525 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.712 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.26.6
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      235456 kB
DirectMap2M:    11298816 kB
DirectMap1G:     7340032 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-09-22/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
fmt check
Diff in /checkout/compiler/rustc_codegen_ssa/src/apple.rs:384:
     /// This is needed by the linker.
     ///
     /// settings["Version"] or settings["VersionMap"]["macOS_iOSMac"][settings["Version"]].
-    pub(crate) fn sdk_version(&self, target: &Target, sdkroot: &Path) -> Result<AppleOSVersion, AppleSdkError> {
+    pub(crate) fn sdk_version(
+        target: &Target,
+        sdkroot: &Path,
+        sdkroot: &Path,
+    ) -> Result<AppleOSVersion, AppleSdkError> {
         if target.abi == "macabi" {
                 .version_map
                 .version_map
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_infer/src/infer/region_constraints/mod.rs" "/checkout/compiler/rustc_infer/src/infer/region_constraints/leak_check.rs" "/checkout/compiler/rustc_infer/src/infer/lexical_region_resolve/mod.rs" "/checkout/compiler/rustc_infer/src/infer/type_variable.rs" "/checkout/compiler/rustc_codegen_ssa/src/base.rs" "/checkout/compiler/rustc_codegen_ssa/src/assert_module_sources.rs" "/checkout/compiler/rustc_codegen_ssa/src/debuginfo/mod.rs" "/checkout/compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs" "/checkout/compiler/rustc_infer/src/infer/canonical/instantiate.rs" "/checkout/compiler/rustc_infer/src/infer/canonical/mod.rs" "/checkout/compiler/rustc_infer/src/infer/canonical/query_response.rs" "/checkout/compiler/rustc_infer/src/infer/canonical/canonicalizer.rs" "/checkout/compiler/rustc_infer/src/infer/opaque_types/mod.rs" "/checkout/compiler/rustc_infer/src/infer/opaque_types/table.rs" "/checkout/compiler/rustc_infer/src/lib.rs" "/checkout/compiler/rustc_codegen_ssa/src/apple/tests.rs" "/checkout/compiler/rustc_codegen_ssa/src/common.rs" "/checkout/compiler/rustc_codegen_ssa/src/mono_item.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/coverageinfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/locals.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/mod.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/rvalue.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/statement.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/analyze.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/intrinsic.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/debuginfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/operand.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/place.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/constant.rs" "/checkout/compiler/rustc_codegen_ssa/src/mir/block.rs" "/checkout/compiler/rustc_codegen_ssa/src/codegen_attrs.rs" "/checkout/compiler/rustc_codegen_ssa/src/size_of_val.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/coverageinfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/misc.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/abi.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/mod.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/type_.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/backend.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/builder.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/declare.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/asm.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/intrinsic.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/debuginfo.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/statics.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/consts.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/write.rs" "/checkout/compiler/rustc_codegen_ssa/src/errors.rs" "/checkout/compiler/rustc_codegen_ssa/src/apple.rs" "/checkout/compiler/rustc_codegen_ssa/src/meth.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/link.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/archive.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/mod.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/rpath/tests.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/metadata.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/rpath.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/command.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/symbol_export.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/lto.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/write.rs" "/checkout/compiler/rustc_codegen_ssa/src/back/linker.rs" "/checkout/compiler/rustc_codegen_ssa/src/target_features.rs" "/checkout/compiler/rustc_codegen_ssa/src/lib.rs" "/checkout/compiler/rustc_mir_transform/src/mentioned_items.rs" "/checkout/compiler/rustc_infer/src/infer/resolve.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Thu Oct 10 01:38:15 UTC 2024
  network time: Thu, 10 Oct 2024 01:38:15 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented:
> - Fix bootstrap so that it doesn't pass SDKROOT by default
>   - Not sure if this is possible, since we also want the user to be able to override SDKROOT when building with bootstrap?

bootstrap can possibly handle that as some kind of target specific option, dependening on the specifics of what you want.

bors commented:
:umbrella: The latest upstream changes (presumably #132497) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
